File: python/penetration_analysis.py
import pandas as pd
import datetime

df = pd.read_excel(r'E:\\Data_analysis_projects\\penetration_analysis\\dataset\\penetration_analysis.xlsx')

df['last_active_date'] = pd.to_datetime(df['last_active_date'])

today_date = pd.to_datetime('2024-09-11')

df['last_active_date_evaluation'] = (today_date - df['last_active_date']).dt.days

df['active_inactive'] = ((df['last_active_date_evaluation'] <= 30) & 
                         (df['sessions'] >= 5) & 
                         (df['listening_hours'] >= 10)).astype(int)

active_users_df = df[df['active_inactive'] == 1]
df_group  = df.groupby('country').agg(
    Active_Users=('active_inactive', 'sum'),
    Total_Users=('user_id', 'count')
).reset_index()

df_group['Active_User_Penetration_Rate'] = (df_group['Active_Users'] / df_group['Total_Users'] * 100).round(2)

print(df_group[['country', 'Active_User_Penetration_Rate']])

File: R/penetration_analysis.R
library(readxl)
library(dplyr)

# Load the dataset
df <- read_excel("E:\\Data_analysis_projects\\penetration_analysis\\dataset\\penetration_analysis.xlsx")

# Convert 'last_active_date' to Date
df$last_active_date <- as.Date(df$last_active_date)

# Define the specific date
today_date <- as.Date("2024-09-11")

# Calculate the difference in days
df <- df %>%
  mutate(last_active_date_evaluation = as.numeric(difftime(today_date, last_active_date, units = "days")))

# Define the criteria for active users
df <- df %>%
  mutate(active_inactive = ifelse(last_active_date_evaluation <= 30 & sessions >= 5 & listening_hours >= 10, "active", "inactive"))

# Calculate the active user penetration rate for each country
penetration_rate <- df %>%
  group_by(country) %>%
  summarise(active_user_penetration_rate = round(mean(active_inactive == "active") * 100, 2))

print(penetration_rate)
